clean_code_fm_guide.md

```diff
@@ -0,0 +1,410 @@
+  case adve_kin{
+    Mons: false
+    Terrain:
+    case adve_kin.ele{
+      VOID    : false
+      FLOOR   : false
+      MON_AREA: true
+    }
+  }
+
+```
+
+
+## **Data Structures**
+### Use getter and setter
+Using getters and setters to access data could be better than simply looking for a property. "Why?" you might ask. Well, here's an unorganized list of reasons why:
+
+- When you want to do more beyond getting an object property, you don't have to look up and change every accessor in your codebase.
+- Makes adding validation simple when doing a set.
+- Encapsulates the internal representation.
+- Easy to add logging and error handling when getting and setting.
+
+**Good:**
+```
+Pos32: Type
+  U32
+
+Pos32.new(x: U32, y: U32, z: U32): Pos32
+  let pos = 0u
+  let pos = U32.or(pos, x)
+  let pos = U32.or(pos, U32.shl(y, 12u))
+  let pos = U32.or(pos, U32.shl(z, 24u))
+  pos
+
+Pos32.get_x(pos: Pos32): U32
+  U32.and(pos, 2047u)
+
+Pos32.set_x(pos: Pos32, x: U32): U32
+  Pos32.new(x, Pos32.get_y(pos), Pos32.get_z(pos))
+```
+
+### Use `new` for datatype with 1 constructor
+**Bad:**
+```
+type Pair <A: Type, B: Type> {
+  pair(fst: A, snd: B)
+}
+
+```
+
+**Good:**
+```
+type Pair <A: Type, B: Type> {
+  new(fst: A, snd: B)
+}
+```
+
+## **Organization**
+### Function callers and callees should be close
+If a function calls another, keep those functions vertically close in the source file. Ideally, keep the caller right above the callee. We tend to read code from top-to-bottom, like a newspaper. Because of this, make your code read that way.
+
+### Identation using 2 spaces
+**Bad:**
+```
+Nat.add(n: Nat, m: Nat): Nat 
+  case n { // No espace
+  zero: m,
+  succ: Nat.succ(Nat.add(n.pred, m)),
+  }
+
+Nat.add(n: Nat, m: Nat): Nat
+    case n { // 4 espaces
+        zero: m,
+        succ: Nat.succ(Nat.add(n.pred, m)),
+    }
+```
+
+**Good:**
+```
+Nat.add(n: Nat, m: Nat): Nat
+  case n { // 2 espaces
+    zero: m,
+    succ: Nat.succ(Nat.add(n.pred, m)),
+  }
+```
+
+### Type notation with reduced spaces
+**Bad:**
+```
+Nat.add( n : Nat, m : Nat) : Nat
+  case n {
+    zero: m,
+    succ: Nat.succ(Nat.add(n.pred , m))
+  }
+
+```
+**Good:**
+```
+Nat.add(n: Nat, m: Nat): Nat
+  case n {
+    zero: m,
+    succ: Nat.succ(Nat.add(n.pred, m))
+  }
+```
+
+### Commas come in the end of the line
+**Bad:**
+```
+Game.move(dir: Dir, pos: Pos32): Pos32
+  case dir {
+    right: Pos32.add(pos, Pos32.new(1u, 0u, 0u))
+    , up: Pos32.sub(pos, Pos32.new(0u, 1u, 0u))
+    , left: Pos32.sub(pos, Pos32.new(1u, 0u, 0u))
+    , down: Pos32.add(pos, Pos32.new(0u, 1u, 0u))
+  }
+```
+**Good:**
+```
+Game.move(dir: Dir, pos: Pos32): Pos32
+  case dir {
+    right: Pos32.add(pos, Pos32.new(1u, 0u, 0u)),
+    up   : Pos32.sub(pos, Pos32.new(0u, 1u, 0u)),
+    left : Pos32.sub(pos, Pos32.new(1u, 0u, 0u)),
+    down : Pos32.add(pos, Pos32.new(0u, 1u, 0u))
+  }
+```
+
+### If possible, align constructors using space
```

Comment:
no!

Reply:
but I'm not sure though, of course the latter looks better, but it is hard to enforce and sometimes it gets ugly. it makes hard to edit the code too